pkg/controller/admissionchecks/multikueue/workload.go

```diff
@@ -0,0 +1,349 @@
+	CreateRemoteObject(ctx context.Context, localClient client.Client, remoteClient client.Client, key types.NamespacedName, workloadName string) error
+	// Copy the status from the job in the worker cluster to the local one.
+	CopyStatusRemoteObject(ctx context.Context, localClient client.Client, remoteClient client.Client, key types.NamespacedName) error
+	// Deletes the Job in the worker cluster.
+	DeleteRemoteObject(ctx context.Context, remoteClient client.Client, key types.NamespacedName) error
+}
+
+type wlGroup struct {
+	local         *kueue.Workload
+	remotes       map[string]*kueue.Workload
+	rc            *remoteController
+	acName        string
+	jobAdapter    jobAdapter
+	controllerKey types.NamespacedName
+}
+
+// the local wl is finished
+func (g *wlGroup) IsFinished() bool {
+	return apimeta.IsStatusConditionTrue(g.local.Status.Conditions, kueue.WorkloadFinished)
+}
+
+// returns true if there is a wl reserving quota
+// the string identifies the remote, ("" - local)
+func (g *wlGroup) FirstReserving() (bool, string) {
+	found := false
+	bestMatch := ""
+	bestTime := time.Now()
+	for remote, wl := range g.remotes {
+		if wl == nil {
+			continue
+		}
+		c := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadQuotaReserved)
+		if c != nil && c.Status == metav1.ConditionTrue && (!found || c.LastTransitionTime.Time.Before(bestTime)) {
+			found = true
+			bestMatch = remote
+			bestTime = c.LastTransitionTime.Time
+		}
+	}
+	return found, bestMatch
+}
+
+func (g *wlGroup) RemoteFinishedCondition() (*metav1.Condition, string) {
+	var bestMatch *metav1.Condition
+	bestMatchRemote := ""
+	for remote, wl := range g.remotes {
+		if wl == nil {
+			continue
+		}
+		if c := apimeta.FindStatusCondition(wl.Status.Conditions, kueue.WorkloadFinished); c != nil && c.Status == metav1.ConditionTrue && (bestMatch == nil || c.LastTransitionTime.Before(&bestMatch.LastTransitionTime)) {
+			bestMatch = c
+			bestMatchRemote = remote
+		}
+	}
+	return bestMatch, bestMatchRemote
+}
+
+func (group *wlGroup) RemoveRemoteObjects(ctx context.Context, cluster string) error {
+	remWl := group.remotes[cluster]
+	if remWl == nil {
+		return nil
+	}
+	if err := group.jobAdapter.DeleteRemoteObject(ctx, group.rc.remoteClients[cluster].client, group.controllerKey); err != nil {
+		return fmt.Errorf("deleting remote controller object: %w", err)
+	}
+
+	if controllerutil.RemoveFinalizer(remWl, kueue.ResourceInUseFinalizerName) {
+		if err := group.rc.remoteClients[cluster].client.Update(ctx, remWl); err != nil {
+			return fmt.Errorf("removing remote workloads finalizeer: %w", err)
+		}
+	}
+
+	err := group.rc.remoteClients[cluster].client.Delete(ctx, remWl)
+	if client.IgnoreNotFound(err) != nil {
+		return fmt.Errorf("deleting remote workload: %w", err)
+	}
+	group.remotes[cluster] = nil
+	return nil
+}
+
+func (a *wlReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
```

Comment:
Indeed, even if we add markers here, manifests aren't updated.
However, we have been adding markers to all reconcilers so that we can clarify which permissions the reconciler must have.

So, I still think that having kubebuilder markers would be better. Or we should add markers to the `batchJobAdapter` instead of here.

Reply:
That could look strange since the adapter is not a reconciler.

Reply:
> That could look strange since the adapter is not a reconciler.

It makes sense. We should put markers here.